src/modules/fancyzones/lib/FancyZones.cpp

```diff
@@ -976,41 +976,42 @@ void FancyZones::UpdateZoneWindows() noexcept
         mi.cbSize = sizeof(mi);
         if (GetMonitorInfo(monitor, &mi))
         {
-            DISPLAY_DEVICE displayDevice = { sizeof(displayDevice) };
-            PCWSTR deviceId = nullptr;
+            DISPLAY_DEVICE displayDevice;
+            displayDevice.cb = sizeof(displayDevice);
 
+            std::wstring deviceId;
             bool validMonitor = true;
-            if (EnumDisplayDevices(mi.szDevice, 0, &displayDevice, 1))
+            DWORD deviceIdx = 0;
+            while (EnumDisplayDevices(mi.szDevice, deviceIdx, &displayDevice, EDD_GET_DEVICE_INTERFACE_NAME))
             {
-                if (WI_IsFlagSet(displayDevice.StateFlags, DISPLAY_DEVICE_MIRRORING_DRIVER))
+                // Only take active monitors (presented as being "on" by the respective GDI view) into account.
+                if (WI_IsFlagSet(displayDevice.StateFlags, DISPLAY_DEVICE_ACTIVE))
                 {
-                    validMonitor = FALSE;
-                }
-                else if (displayDevice.DeviceID[0] != L'\0')
-                {
-                    deviceId = displayDevice.DeviceID;
+                    validMonitor = WI_IsFlagClear(displayDevice.StateFlags, DISPLAY_DEVICE_MIRRORING_DRIVER);
+                    deviceId = std::wstring(displayDevice.DeviceID);

```

Comment:
I will address some of suggestions that You have, but as for the 

> this code also doesn't support having multiple monitors per adapter, since it always picks the first/last one

that is something I've discussed with @enricogior and we agreed to go with last active (non-mirroring) display in the list, now I can change this if we all agree that should be done (or maybe do it in the separate PR)

Reply:
I suspect we need to support multiple active monitors per device, otherwise it might cause regressions.
We also need to verify those active monitors are not duplicated.
For example in the case of the Surface docking station, the internal monitor might be active for each device and we don't want to add it multiple times.

Reply:
OK, I'll implement approach @yuyoyuppe suggested.